Close incoming HTTP Connection if requested

sgothel opened the pull request:
* Resolves: #10042
* Target version: master 

### Summary

cool#10042: Close incoming HTTP Connection if requested (Regression of cool#9621)
    
Before commit 8c05bdcd74009eed11be2c247dfd8a5fd6c1c9fa
for cool#9621 `Keep Connection Open for files served via FileServerRequestHandler and ClientRequestDispatcher ...`,
all client connections got closed.
    
Since this change, we try keeping all connections open.
However, an explicit client request w/ `Connection: close` got ignored
in `ClientRequestDispatcher::handleIncomingMessage`.
    
This patch reverts behavior as follows:
- If the request has been handled synchronously (served==true),
  then socket->eraseFirstInputBytes(body-bytes)

- If close-connection is requested via HTTP/1.1 header `Connection: close`
  or if HTTP/1.0 header does not contain `Connection: keep-alive`
  _and_ the request has been handled synchronously (served==true), then:
  - Close connection if requested via HTTP Header
  - Set HTTP/1.1 reply header `Connection: close`

This patch addresses cool#9621, commit 895c224efae9c21f0481e2fbf024a015656a5a97 and cool#10042.
    
Tested open and closed connection requests via UnitSession.cpp
as well as with pre-existing tests like integration-http-server.cpp.

### Checklist

- [X] I have run `make prettier-write` and formatted the code.
- [X] All commits have Change-Id
- [X] I have run tests with `make check`
- [X] I have issued `make run` and manually verified that everything looks okay
- [X] Documentation (manuals or wiki) has been updated or is not required

sgothel commented:
Note: CodeQL claims
```
/home/runner/work/online/include/LibreOfficeKit/LibreOfficeKit.hxx:919:10: note: candidate: ‘bool lok::Document::createSlideRenderer(int, unsigned int*, unsigned int*, bool, bool)’
  919 |     bool createSlideRenderer(
      |          ^~~~~~~~~~~~~~~~~~~
/home/runner/work/online/include/LibreOfficeKit/LibreOfficeKit.hxx:919:10: note:   candidate expects 5 arguments, 6 provided
```

I updated LO today to commit 9031c056715fe43525d44704c05a1380b68eb278, which provides the 6 arguments method.

sgothel commented:
Note: `unit-wopi-async-slow` passed on the Cypress tests as well locally, but probably not stable as shown in the unit tests w/o Cypress.

mmeeks commented on wsd/ClientRequestDispatcher.cpp:
```diff
@@ -926,18 +928,36 @@ void ClientRequestDispatcher::handleIncomingMessage(SocketDisposition& dispositi
-        socket->eraseFirstInputBytes(map._messageSize - map._headerSize);
+        // If we succeeded - remove the request header from our input buffer
+        socket->eraseFirstInputBytes(map._headerSize);
+        // Some handlers, like handleClientProxyRequest, used in the richdocumentscode
```
Looks really good - but I'd prefer not to have such big comments in-line in the code - but have them in the unit test and/or git commit log - with perhaps a public github ticket reference here for documentation if necessary in future.

sgothel replied:
will rename served -> servedSync and add the notion to it. This patch clarifies this whole commend AFAIK and hence renders it redundant. Will add a ref to the issues, yup.

mmeeks reviewed (approved):
Looks great to me as/when it passes CI - and @timar we should get this into our stable / release branches ASAP etc. but - I'd -really- want to see a unit test for this.
Good to fix this properly - satisfying =)

sgothel commented:
> Looks great to me as/when it passes CI - and @timar we should get this into our stable / release branches ASAP etc. but - I'd -really- want to see a unit test for this. Good to fix this properly - satisfying =)

forced pushed ..

Added notion of 'This patch addresses cool#9621, commit 895c224efae9c21f0481e2fbf024a015656a5a97 and cool#10042.'
    
Tested open and closed **the added test** connection requests via UnitSession.cpp
as well as with pre-existing tests like integration-http-server.cpp.

+++

Regarding CI passing .. well, it all passed here locally and I added some notes re CodeQL(C++) .. the usual I am afraid.

mmeeks commented:
Just merged Ash' unit test improvements - perhaps if you re-base on that life will be better ? =)

sgothel commented:
> Just merged Ash' unit test improvements - perhaps if you re-base on that life will be better ? =)

Was `unit-wopi-async-slow` last time, now its `unit-wopi-fail-upload` .. while locally all of them passed here,
plus on other test machines above.

hfiguiere commented:
works fine here. Thanks !

caolanm commented:
backport to 24.04 as https://github.com/CollaboraOnline/online/pull/10063
